### DISCUSSION TITLE
Break a fori_loop

### DISCUSSION THREAD

**Discussion by aespielberg** [jax-ml/jax#8375] [Q&A] [answered]
Is there any way to break out of a fori_loop early?  I understand why jax would need to precompile a maximum number of loop iterations, but is there any way to have it run for only a subset of those loop iterations at runtime?

**Comment by jakevdp** [comment, marked as answer]
There's not any way to break a `fori_loop`; if you need to break based on some condition you can do so using a `while_loop`. It might look something like this:
```python
from jax import lax

def func(i, x):
  return x + i
print(lax.fori_loop(lower=0, upper=10, body_fun=func, init_val=0))
# 45

def body_fun(carry):
  i, x = carry
  return i + 1, func(i, x)

def cond_fun(carry):
  i, x = carry
  break_condition = (i == 20)
  return ~break_condition & (i < 10)

print(lax.while_loop(cond_fun, body_fun, init_val=(0, 0))[1])
# 45
```

**Reply by aespielberg** [reply]
Ah, that's good to know, but then I lose reverse-mode differentiability.  It would be really nice to have that but I guess that's not available right now.  I read in the docs that it's not doable because jax needs to know how much memory to allocate, but I feel like it should be possible to allocate a maximum amount of memory to make it work.

**Reply by jakevdp** [reply]
I see. Well, perhaps in lieu of breaking you could have some break condition such that your `body_fun` stops updating the output once it's met?

**Reply by aespielberg** [reply]
Was just having this discussion with a colleague!  Yes, I will try that by putting a jnp.where in the for loop.  I'm not sure how the compiler will optimize that - I don't want it to evaluate both branches if it doesn't have to.  Do you happen to know?  (of course I will attempt to try it soon)

**Reply by mattjj** [reply]
I think `jnp.where` will evaluate both branches. (Technically the compiler has freedom not to evaluate both sides, but in practice it'll evaluate both.)

You can use `jax.lax.cond` to avoid evaluating both sides. It's another slightly awkward control flow construct though. And on GPU these things can have significant kernel launch overheads.

**Reply by mattjj** [reply]
By the way, this need comes up periodically (to have an early-exit-but-differentiable-bounded-iteration-loop), but while it's doable it's not trivial so it's never made it to the top of our priority list. It'd make a good starter project for someone, except it requires learning a bunch of internal machinery first...

Let us know if the fori_loop + cond works for you, or otherwise if this becomes a blocker for your work.

**Reply by aespielberg** [reply]
FWIW, I am pretty sure that jnp.cond is actually evaluating both branches in a different context (which is now becoming a bottleneck for this research project as it is using a tooooon of memory): https://github.com/google/jax/issues/8409

**Reply by shailesh1729** [reply]
Kindly allow me to digress a little bit. Differentiability is not a key concern in many numerical linear algebra routines which aim to solve problems like  A x = b for underdetermined systems, computing a few singular values for truncated SVD based algorithms, etc. While I can code such algorithms using JAX but static size restrictions, lack of early exit in `fori_loop`, requiring that the shape of algorithm state in `fori_loop` or `while_loop` iterations cannot change, means that their natural implementation cannot be put under `jit`.  And without `jit`, JAX is usually quite slow.  One has to go through hackish ways to make them work under `jit`. But it doesn't scale well.  

Consider a [singular value thresholding](https://statweb.stanford.edu/~candes/software/svt/) algorithm for matrix completion problem.  I may have a huge matrix `X` which is quite sparse (I can now store `X` efficiently using `jax.experimental.sparse`). One way is to compute dense SVD (supported by JAX) and then do thresholding. But this is wasteful if X is of the order of 100s of thousands but the actual singular values of concerns are only the top few hundred. My ideal would be to compute a small (say r) number of singular values each time, which may require a variable number of bidiagonalization steps to converge and stop when enough singular values have stabilized up to the specified threshold. Such an algorithm is next to impossible to code and compile under the current `jit`. I would have to do a worst-case estimate for the number of iterations required for the algorithm to converge so that I can allocate space for storing the intermediate Ritz vectors and other variables as part of the algorithm state.  I had recently coded a jittable truncated SVD routine using JAX but I found that it is actually running slower on a GPU than the reference implementation in MATLAB running on a CPU. It is still a significant improvement from full SVD though for huge+sparse matrices. 

I do understand that `jit`, `grad`, `pmap`, `vmap`, `jvp`, primitives are designed to work interchangeably which imposes these restrictions. But a separate version of `jit` which doesn't require differentiability on the input function, but only focuses on generating efficient code for the underlying hardware, may open up JAX to many other areas. That special `jit` could potentially allow dynamic output shapes (depending on function arguments or early stopping) as well as dynamic slices in the intermediate steps. 

Please forgive me if some of these comments seem immature as I am still quite new to JAX.

**Reply by aespielberg** [reply]
Well, I should note that i do care about differentiability, and I think adding a second jit is probably a huge task.  I think that there are a ton of hidden catches in what you are proposing and I don't think most of this is related to the thread at hand.  If you want to propose jit2, I think you should open a feature request.  

What I am asking about is something that should be completely feasible in terms of differentiability and the current jit framework (thought maybe with a bit of work), assuming both `jax.lax.cond` already works, and given the fact that a differentiable for loop is essentially composing the same function that only has to be composted into itself repeatedly with an input `jvp` for backwards diff.

**Reply by patrick-kidger** [reply]
For those following this topic / those who come across this topic.

[Diffrax](https://github.com/patrick-kidger/diffrax) (a pure-JAX library of ODE/SDE/etc. solvers) implements such an "early-exit-but-differentiable-bounded-iteration-loop" as `bounded_while_loop`, here:
https://github.com/patrick-kidger/diffrax/blob/2b4e4d863c15abc7143919bac7825090bbfe50be/diffrax/misc/bounded_while_loop.py

For those curious about the technical details and caveats, have a look at my reply here:
https://github.com/google/jax/issues/2139#issuecomment-1039293633

**Reply by aespielberg** [reply]
Awesome, thank you so much!  I will try this out soon!

**Comment by patrick-kidger** [comment]
Sorry, meant to post as a proper answer, not just in the comments:

https://github.com/google/jax/discussions/8375#discussioncomment-2174547

[TL;DR links and explanation for an implementation of a reverse-mode differentiable bounded while loop.]

**Reply by njwfish** [reply]
Is there any possibility of building this into base jax? It is extremely useful!
